PyTorch-Sparse-Roadmap.md

```diff
@@ -0,0 +1,440 @@
+
+## Related Work
+
+Much prior work on sparse matrices is implemented in the 
+[`scipy.sparse`](https://docs.scipy.org/doc/scipy/reference/sparse.html) 
+module; the [`pydata/sparse`](https://github.com/pydata/sparse) project extends 
+some of those ideas to sparse arrays of generic rank `n`. As a stretch goal, it 
+may make sense to have general rank `n` PyTorch tensors being supported in most 
+of the same ways as `pydata/sparse` and to have sparse tensors of rank 2 (or 
+with `sparse_dim()==2`) supported in most of the same ways as `scipy.sparse`. 
+
+Among projects tailored to neural networks & deep learning,
+[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul)
+has a `SparseTensor` class that uses COO format for tensors of arbitrary
+rank `n` and a `SparseMatrix` class that uses CSR format for tensors of
+rank 2; explicit conversion routines between formats exist where they apply. 
+There are also rank 2 `CSRNDArray` and rank 1 `RowSparseMatrix` classes 
+included in
+[MXNet](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul).
+The project [`pytorch_sparse`](https://github.com/rusty1s/pytorch_sparse)
+suggests some directions for supporting efficient sparse matrix operations
+with `autograd` and also with CUDA support.
+
+Finally, there are generalizations of sparse matrix storage formats to sparse 
+tensors of arbitrary rank `n` described in various articles listed in the 
+references.
+
+## Implementation
+
+The module `torch` already contains an attribute `torch.sparse_coo` of type
+`torch.layout` with an associated constructor `torch.sparse_coo_tensor`. The
+proposed CSR implementation should extend the API by adding a few attributes
+to the `torch` module with similar names:
+
++ `torch.sparse_csr2d` (an attribute of type `torch.layout`);
++ `torch.sparse_csr2d_tensor` (an associated constructor); and
++ functions `torch.to_sparse_csr2d` & `torch.to_sparse_coo` for explicit
+  conversion between representations (in the case `sparse_dims()==2`).
+
+The use of `_csr2d` in the new identifiers emphasizes the fact that the CSR
+format applies only to tensors of rank 2.
+
+The basic outline, then, is as follows:
+
+1. Construct an exhaustive suite of benchmark tests to time the existing
+   implementation of sparse matrix-vector & matrix-matrix products in PyTorch.
+
+   + See [issue # 16187](https://github.com/pytorch/pytorch/issues/16187)
+     for representative examples of benchmarks.
+   + Include examples of matrices of fairly large dimensions (i.e., where dense
+     representations are impractical or infeasible).
+   + Include examples with `sparse x dense` and `sparse x sparse` matrices for
+     the timings.
+
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr2d` and `torch.to_sparse_coo` to convert sparse PyTorch
+   tensors of sparse rank 2 between COO and CSR formats.
```

Comment:
If CSR is going to be another layout type, there is some infrastructural work that is going to need to be done here; you need a new tensor type (since a CSR isn't going to have the same memory representation as COO tensor). For example, right now we refer to COO sparse as just "Sparse" in the codebase; it might make sense to disambiguate these names to "SparseCOO" so people don't mix them up. (Or, I suppose, we can just grandfather sparse to mean COO sparse, and be explicit in all the new locations).

Reply:
The identity "Sparce == Sparce COO" is not consistent in current pytorch, for instance, there are sparse tools that include `coo` explicitly in their names: `torch.sparse_coo_tensor`, `at::_sparse_coo_tensor_with_dims`, `torch.sparse_coo`, etc.
This indicates that the design of the current infrastructure implicitly would support other sparse formats. For instance, one could have  `torch.sparse_csr_tensor`, etc.

Reply:
I think Pearu's comment is correct; it is possible to extend the current sparse classes without having to specify the storage format by a name like "SparseCOO." This could perhaps be achieved by adding a `_layout` attribute of type `torch.sparse_coo` or `torch.sparse_csr` or some other valid PyTorch object of class `torch.layout`.

The proposal has been modified to provide a more thorough description of the CSR format and the associated constructor.